cgrok/grok.go

```diff
@@ -201,39 +218,42 @@ func (match *Match) Captures() map[string][]string {
 	return captures
 }
 
-func (match *Match) Free() {
-	ptr := unsafe.Pointer(match.gm.subject)
-	if uintptr(ptr) != 0 {
-		C.free(ptr)
-	}
-	C.grok_match_free(&match.gm)
-}
-
 /* Add captures to the map, clobbering existing keys. Only extracts
    sub-expressions which have been explicitly renamed - `%{DAY}` will not be extracted,
    but `%{DAY:day}` will be extracted with the key `day`. Sub-expressions with duplicate names
    will clobber each other as well - the last match will remain. */
 func (match *Match) CaptureIntoMap(captures map[string]string) {
-	var name, substring *C.char
-	var namelen, sublen C.int
+	var name *C.char
+	var namelen, suboffset, sublen C.int
 
 	C.grok_match_walk_init(&match.gm)
 
-	for C.grok_match_walk_next(&match.gm, &name, &namelen, &substring, &sublen) == GROK_OK {
-		gname := C.GoStringN(name, namelen)
-		if idx := strings.Index(gname, ":"); idx > -1 {
-			gname = gname[(idx+1):]
-		} else {
-			// The name doesn't include a colon, skip this group
-			continue
+	for C.grok_match_walk_next_offsets(&match.gm, &name, &namelen, &suboffset, &sublen) == GROK_OK {
+		gname := match.grok.gostringn(name, namelen)
+		colonIdx := strings.Index(gname, ":")
+		if colonIdx > -1 {
+			/* Do some pointer arithmetic to find the index of the match within the string. The match should always
+                           be between the beginning of the subject and the end, or an empty string */
+			var gsubstring string
+			if int(sublen) > 0 {
+				if int(suboffset) > -1 && int(suboffset + sublen) <= len(match.subject) {
```

Comment:
1) This is a good question. In practice on x84_64 `int` is always 64 bits anyways, and this is a pretty common pattern so I want to say Go's `int` should at least as big as the C `int` size. But if we move to non-x86_64 platforms we may want to revisit that assumption.

2) Yes - we have to allow matches that extend to the imaginary position just after the last character. Imagine a 2 character-long string, "ab". If the whole string matches, the match would be `suboffset = 0; sublen = 2`, so `suboffset + sublen == 2` and `len(match.subject) == 2`.